### PR TITLE
Remove blocking calls and change threat intel feed flow to event driven

### LINKED ISSUE #870: [BUG] Detector creation gets stuck on clusters with large shards and heavy ingestion
**What is the bug?**
```
curl localhost:9200/_cat/tasks?v | less
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
action                                                     task_id                        parent_task_id                 type      start_time    timestamp running_time ip            node
cluster:admin/opensearch/securityanalytics/detector/write  NS5L3EYoSM2ED7Ivqq1snQ:990205  -                              transport 1706578051254 01:27:31  5.2h         10.212.107.51 5112ba5b511cfd4495
cluster:admin/opensearch/securityanalytics/detector/write  NS5L3EYoSM2ED7Ivqq1snQ:991197  -                              transport 1706578171258 01:29:31  5.2h         10.212.107.51 5112ba5b511cfd4495
cluster:admin/opensearch/securityanalytics/rule/search     wWwwf7eRSD2oo8KulgOF7Q:917083  -                              transport 1706578176167 01:29:36  5.2h         10.212.27.178 a173576e5c9b149d2e
cluster:admin/opendistro/alerting/monitor/write            NS5L3EYoSM2ED7Ivqq1snQ:991834  -                              transport 1706578242304 01:30:42  5.2h         10.212.107.51 5112ba5b511cfd4495
cluster:admin/opensearch/securityanalytics/detector/write  6x4YBILlRNqCh-H5SEGz4g:929275  -                              transport 1706578277667 01:31:17  5.2h         10.212.98.228 f8a85ed4b86db333fc
cluster:admin/opensearch/securityanalytics/mapping/get     NS5L3EYoSM2ED7Ivqq1snQ:992971  -                              transport 1706578360527 01:32:40  5.1h         10.212.107.51 5112ba5b511cfd4495
indices:admin/mappings/get   
```

**How can one reproduce the bug?**

There are a few blocking calls (due to invocation of `actionGet()`) which are causing deadlocks in detector creation flow. On clusters with heavy ingestion and large shards this problem is magnified and causes cluster to choke up and run out of resources stuck in deadlocks
-
**What is the expected behavior?**
Code should be event driven using the listener-based SPIs exposed by opensearch transport client

### PR DESCRIPTION
**Pull request by eirsep**
### Description
Remove blocking calls and change threat intel feed flow to event driven 

### Issues Resolved
#870

### CODE REVIEW

**Review by engechas — COMMENTED** [review]
General comment on removal of the try/catches. Rest LGTM

File: src/main/java/org/opensearch/securityanalytics/threatIntel/DetectorThreatIntelService.java [hunk 1]
```diff
@@ -116,13 +116,10 @@ private String buildQueryStringQueryWithIocList(Set<String> iocs) {
      * Fetches threat intel data and creates doc level queries from threat intel data
      */
     public void createDocLevelQueryFromThreatIntel(List<LogType.IocFields> iocFieldList, Detector detector, ActionListener<List<DocLevelQuery>> listener) {
-        try {
```
[the remark this answers is not in the document]
Comment (eirsep) [reply src/main/java/org/opensearch/securityanalytics/threatIntel/DetectorThreatIntelService.java:119]: listener framework is event driven and no catch is required as `ActionListener.onFailure()` would need to implement whatever logic was written in catch block as callback mechanism will not throw an exception

File: src/main/java/org/opensearch/securityanalytics/threatIntel/ThreatIntelFeedDataService.java [hunk 2]
```diff
@@ -103,24 +102,11 @@ public ThreatIntelFeedDataService(
     public void getThreatIntelFeedData(
             ActionListener<List<ThreatIntelFeedData>> listener
     ) {
-        try {
```
[the remark this answers is not in the document]
Comment (eirsep) [reply src/main/java/org/opensearch/securityanalytics/threatIntel/ThreatIntelFeedDataService.java:106]: listener framework is event driven and no catch is required as `ActionListener.onFailure()` would need to implement whatever logic was written in catch block as callback mechanism will not throw an exception

File: src/main/java/org/opensearch/securityanalytics/threatIntel/DetectorThreatIntelService.java [hunk 3]
```diff
@@ -116,13 +116,10 @@ private String buildQueryStringQueryWithIocList(Set<String> iocs) {
...
             if (false == detector.getThreatIntelEnabled() || iocFieldList.isEmpty()) {
                 listener.onResponse(Collections.emptyList());
                 return;
             }
-
-            CountDownLatch latch = new CountDownLatch(1);
```
[the remark this answers is not in the document]
Comment (eirsep) [reply src/main/java/org/opensearch/securityanalytics/threatIntel/DetectorThreatIntelService.java:125]: bad practice. The right construct to use is a `Countdown` 
```
Countdown is a  simple thread safe count-down class that in contrast to a CountDownLatch never blocks. This class is useful if a certain action has to wait for N concurrent tasks to return or a timeout to occur in order to proceed.
```
but safer to do it the event-driven way

File: src/main/java/org/opensearch/securityanalytics/threatIntel/ThreatIntelFeedDataService.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (engechas) [reply src/main/java/org/opensearch/securityanalytics/threatIntel/ThreatIntelFeedDataService.java:106]: If this code throws an exception, we never make a call to `ActionListener.onFailure()`. I believe the original call just hangs in this case, at least that is what I have experienced while developing in this package.

[the remark this answers is not in the document]
Comment (eirsep) [reply src/main/java/org/opensearch/securityanalytics/threatIntel/ThreatIntelFeedDataService.java:106]: ok 
reverting try-catch removal

File: src/main/java/org/opensearch/securityanalytics/threatIntel/jobscheduler/TIFJobParameterService.java [hunk 4]
```diff
@@ -104,82 +106,72 @@ private String getIndexMapping() {
...
     /**
      * Update jobSchedulerParameter in an index {@code TIFJobExtension.JOB_INDEX_NAME}
+     *
      * @param jobSchedulerParameter the jobSchedulerParameter
      */
     public void updateJobSchedulerParameter(final TIFJobParameter jobSchedulerParameter, final ActionListener<ThreatIntelIndicesResponse> listener) {
         jobSchedulerParameter.setLastUpdateTime(Instant.now());
         StashedThreadContext.run(client, () -> {
             try {
-                if (listener != null) {
```
[the remark this answers is not in the document]
Comment (eirsep) [reply src/main/java/org/opensearch/securityanalytics/threatIntel/jobscheduler/TIFJobParameterService.java:113]: changed it to non-null in all invocations

File: src/main/java/org/opensearch/securityanalytics/transport/TransportIndexDetectorAction.java [hunk 5]
```diff
@@ -716,8 +716,6 @@ private IndexMonitorRequest createDocLevelMonitorRequest(List<Pair<String, Rule>
     }
 
     private void addThreatIntelBasedDocLevelQueries(Detector detector, ActionListener<List<DocLevelQuery>> listener) {
-        try {
```
Comment (engechas) [inline src/main/java/org/opensearch/securityanalytics/transport/TransportIndexDetectorAction.java:719]: Try/catch still removed here

**Review by engechas — APPROVED** [review]

**Review by jowg-amazon — APPROVED** [review]

**Pull request merged**